File: frontend/lib/services/functions/FolderService.dart
import 'dart:convert';
import 'dart:developer';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:frontend/services/functions/AuthService.dart';
import 'package:frontend/services/models/course.dart';
import 'package:frontend/services/models/folder.dart';
import 'package:frontend/utils/constants.dart';
import 'package:http/http.dart' as http;

class FolderService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  String idToken = AuthService.idToken!;

  final CollectionReference folders =
      FirebaseFirestore.instance.collection('folders');

  Stream<QuerySnapshot> getFolderStreamByUser(String userId) {
    final foldersStream =
        folders.where('userId', isEqualTo: userId).snapshots();
    return foldersStream;
  }

  Future<void> updateFolder(
      String folderId, Map<String, dynamic> updatedData) async {
    try {
      await folders.doc(folderId).update(updatedData);
      log('Folder updated successfully');
    } catch (e) {
      log('Error updating folder: $e');
    }
  }

  Future<void> createFolder(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse(AppConstants.FOLDER_API),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
      body: jsonEncode(data),
    );

    if (response.statusCode != 201) {
      throw Exception('Failed to create folder');
    }
  }

  //get folder from id
  Future<Folder> getFolder(String folderId) async {
    final response = await http.get(
      Uri.parse("${AppConstants.FOLDER_API}/$folderId"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
    );

    if (response.statusCode == 200) {
      return Folder.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to get folder');
    }
  }

  //delete folder
  Future<void> deleteFolder(String folderId) async {
    final response = await http.delete(
      Uri.parse("${AppConstants.FOLDER_API}/$folderId"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to delete folder');
    }
  }

  Future<List<Folder>> getFoldersOfUser(String userId) async {
    final response = await http.get(
      Uri.parse("${AppConstants.FOLDER_API}/userFolders?userId=$userId"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data
          .map((json) => Folder.fromJson(json as Map<String, dynamic>))
          .toList();
    } else {
      throw Exception('Failed to get user folders');
    }
  }

  Future<void> addCourseToFolder(String folderId, String courseId) async {
    final response = await http.post(
      Uri.parse("${AppConstants.FOLDER_API}/addCourseToFolder"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
      body: json.encode({'folderId': folderId, 'courseId': courseId}),
    );

    if (response.statusCode != 200) {
      log('Response body: ${response.body}');
      throw Exception('Failed to save course');
    }
  }

  Future<List<dynamic>> getCoursesIdFromFolder(String folderId) async {
    final response = await http.get(
      Uri.parse("${AppConstants.FOLDER_API}/$folderId/courses"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
    );
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.toList();
    } else {
      throw Exception('Failed to get courses');
    }
  }

  Future<void> removeCourseFromFolder(String folderId, String courseId) {
    return http
        .delete(
      Uri.parse("${AppConstants.FOLDER_API}/$folderId/removeCourse/$courseId"),
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
    )
        .then((response) {
      if (response.statusCode != 200) {
        throw Exception('Failed to remove course');
      }
    });
  }
}

File: backend/src/app/service/FolderService.js
const { FolderCollection } = require('./Collections');
const { firestore } = require('firebase-admin');

exports.getFoldersofUser = async (userId) => {
    try {
        const snapshot = await FolderCollection
            .where('userId', '==', userId)
            .get();

        if (snapshot.empty) {
            return [];
        }

        const folders = [];
        snapshot.forEach(doc => {
            folders.push({ id: doc.id, ...doc.data() });
        });

        return folders;
    } catch (error) {
        console.error(`Error getting folders of user ${userId}:`, error);
        throw error;
    }
};

exports.createFolder = async (data) => {
    try {
        const docRef = FolderCollection.doc();
        await docRef.set({
            ...data,
            courses: []
        });
        return { "folderId": docRef.id };
    } catch (error) {
        console.error('Error creating folder:', error);
        throw error;
    }
};

exports.addCourseToFolder = async (folderId, courseId) => {
    try {
        const docRef = FolderCollection.doc(folderId);
        await docRef.update({
            courses: firestore.FieldValue.arrayUnion(courseId)
        });
        return { "message": "courseId added successfully" };
    } catch (error) {
        console.error('Error adding courseId to folder:', error);
        throw error;
    }
};

exports.deleteFolder = async (folderId) => {
    try {
        const docRef = FolderCollection.doc(folderId);
        await docRef.delete();
        return { "message": "Folder deleted successfully" };
    } catch (error) {
        console.error('Error deleting folder:', error);
        throw error;
    }   
}

exports.getFolderById = async (folderId) => {
    try {
        const doc = await FolderCollection.doc(folderId).get();
        if (!doc.exists) {
            return null;
        }

        return { id: doc.id, ...doc.data() };
    } catch (error) {
        console.error(`Error getting folder ${folderId}:`, error);
        throw error;
    }
}

exports.getCoursesOfFolder = async (folderId) => {
    try {
        const doc = await FolderCollection.doc(folderId).get();
        if (!doc.exists) {
            return null;
        }

        return doc.data().courses;
    } catch (error) {
        console.error(`Error getting courses of folder ${folderId}:`, error);
        throw error;
    }
}